[EarlyCSE] Compare GEP instructions based on offset

dianqk opened the pull request:
Closes #65763.
This will provide more opportunities for constant propagation for subsequent optimizations.

dianqk committed:
[EarlyCSE] Pre-commit offset-based GEP (NFC)

dianqk committed:
[EarlyCSE] Add a vectorization failure example (NFC)

nikic commented:
Please check the compile-time impact of this change. The impact was high when I tried something like this before, but maybe I made a mistake.

DianQK commented:
> Please check the compile-time impact of this change. The impact was high when I tried something like this before, but maybe I made a mistake.

Could you add https://github.com/DianQK/llvm-project to https://llvm-compile-time-tracker.com/? Thanks.

nikic commented:
> > Please check the compile-time impact of this change. The impact was high when I tried something like this before, but maybe I made a mistake.
> 
> Could you add https://github.com/DianQK/llvm-project to https://llvm-compile-time-tracker.com/? Thanks.

I'll do this on Thursday. I'm currently on vacation and don't have the SSH key with me.

nikic commented on llvm/lib/Transforms/Scalar/EarlyCSE.cpp:
```diff
@@ -336,6 +336,13 @@ static unsigned getHashValueImpl(SimpleValue Val) {
+    auto &DL = GEP->getModule()->getDataLayout();
+    APInt Offset(DL.getIndexTypeSizeInBits(GEP->getType()), 0);
+    if (GEP->accumulateConstantOffset(DL, Offset))
+      return hash_combine(GEP->getOpcode(), GEP->getOperand(0), Offset);
```
nit: getPointerOperand().

nikic commented on llvm/lib/Transforms/Scalar/EarlyCSE.cpp:
```diff
@@ -405,6 +412,17 @@ static bool isEqualImpl(SimpleValue LHS, SimpleValue RHS) {
+      return false;
+    auto &DL = LGEP->getModule()->getDataLayout();
+    APInt LOffset(DL.getIndexTypeSizeInBits(LGEP->getType()), 0);
+    APInt ROffset(DL.getIndexTypeSizeInBits(LGEP->getType()), 0);
```
nit: Avoid duplicate getIndexTypeSizeInBits() call (can't have different bitwidths if the pointer operand is the same)

DianQK commented:
> > > > Please check the compile-time impact of this change. The impact was high when I tried something like this before, but maybe I made a mistake.
> > > 
> > > 
> > > Could you add https://github.com/DianQK/llvm-project to https://llvm-compile-time-tracker.com/? Thanks.
> > 
> > 
> > I'll do this on Thursday. I'm currently on vacation and don't have the SSH key with me.
> 
> Done, results are here: http://llvm-compile-time-tracker.com/compare.php?from=ac1daad9bb4eb083df6b215c029816d3149e00d8&to=e079f3f9107f55e1595a47aafec5fe4e03277665&stat=instructions:u

Sad result. My initial guess was that in `GEP 1 1 .... 1 1 a` did redundant calculations.
But https://github.com/DianQK/llvm-project/commit/f3fdea44cd6ba0b129337850fecf567c35de6790 got [worse results](https://llvm-compile-time-tracker.com/?config=NewPM-O3&stat=instructions%3Au&remote=DianQK).
My other guess is that iteration itself is expensive. I'll try again. If that still doesn't work, I'd look into valgrind related tools.

dianqk committed:
[EarlyCSE] Compare GEP instructions based on offset

This will provide more opportunities for
constant propagation for subsequent optimizations.

DianQK commented:
It's not ready for review yet. It may contain many nits.

Progress update:
Since lookups require frequent computation of constant offsets. GEP should not be considered as a `SimpleValue`.
I got better results than the last commit. I'll analyze it next with specific source code.

https://llvm-compile-time-tracker.com/compare.php?from=ac1daad9bb4eb083df6b215c029816d3149e00d8&to=31e2ec9d89aeded9ea5da822262449b0c4e8ab16&stat=instructions:u

nikic commented on llvm/lib/Transforms/Scalar/EarlyCSE.cpp:
```diff
@@ -553,6 +552,77 @@ bool DenseMapInfo<CallValue>::isEqual(CallValue LHS, CallValue RHS) {
+
+} // end namespace llvm
+
+unsigned DenseMapInfo<GEPValue>::getHashValue(GEPValue Val) {
```
```suggestion
unsigned DenseMapInfo<GEPValue>::getHashValue(const GEPValue &Val) {
```
Here and elsewhere.

dianqk committed:
[EarlyCSE] Use ref

dianqk committed:
fixup! [EarlyCSE] Compare GEP instructions based on offset

Add comments and some minor changes

DianQK commented:
Using references works well!
https://llvm-compile-time-tracker.com/compare.php?from=ac1daad9bb4eb083df6b215c029816d3149e00d8&to=089349f9eb47c57cfa2049ce0af5c7dd079a8581&stat=instructions%3Au

I've also tried some approaches. 
It should be useful to avoid offset calculations. I tried counting only with a GEP i8, but there is some extra cost to the determination of this one.
Simplifying the offset calculation is of some use, but I'm not sure it makes sense. Results are here https://llvm-compile-time-tracker.com/compare.php?from=ac1daad9bb4eb083df6b215c029816d3149e00d8&to=8ba478eb77c6dc09b87e9b511614e45e167be2b5&stat=instructions%3Au

nikic commented:
Okay, those results look reasonable. It's somewhat unfortunate that this requires duplicating a bunch of code for the separate GEP handling. I'm wondering, would we be able to achieve similar results if we add a `DenseMap<GetElementPtrInst *, APInt>` cache for the offset but otherwise keep the previous code structure? Or would the extra hash lookup make this slow again?

DianQK commented:
> I'm wondering, would we be able to achieve similar results if we add a `DenseMap<GetElementPtrInst *, APInt>` cache for the offset but otherwise keep the previous code structure? Or would the extra hash lookup make this slow again?

I feel like I misunderstood. Do you mean putting the GEP back into `SimpleValue`? And then create a global (or EarlyCSE only) cache value?

nikic commented on llvm/lib/Transforms/Scalar/EarlyCSE.cpp:
```diff
@@ -553,6 +553,77 @@ bool DenseMapInfo<CallValue>::isEqual(CallValue LHS, CallValue RHS) {
+struct GEPValue {
+  Instruction *Inst;
+  APInt ConstantOffset;
+  bool HasConstantOffset;
```
Use std::optional instead?

DianQK replied:
I seem to get a poor performance. Maybe I missed something, I'll look at it again.
- Compare it to using bool: https://llvm-compile-time-tracker.com/compare.php?from=12eda5cf2df6438453683ce8c917a64786686a69&to=ff70ae265ac5ad9da12836ae954fd64ce5b38df4&stat=instructions:u
- Full: https://llvm-compile-time-tracker.com/compare.php?from=ac1daad9bb4eb083df6b215c029816d3149e00d8&to=ff70ae265ac5ad9da12836ae954fd64ce5b38df4&stat=instructions%3Au

nikic commented on llvm/lib/Transforms/Scalar/EarlyCSE.cpp:
```diff
@@ -553,6 +553,77 @@ bool DenseMapInfo<CallValue>::isEqual(CallValue LHS, CallValue RHS) {
+  }
+
+  static bool canHandle(Instruction *Inst) {
+      return isa<GetElementPtrInst>(Inst);
```
The indents in these functions look wrong -- clang-format?

DianQK replied:
https://github.com/llvm/llvm-project/pull/65875/files#diff-77f80fb318742b6bfa0825b341fb6e8cf89141fe8200f3574d556aa48db5114fL551-R551
This line breaks git-based incremental formatting.

nikic commented on llvm/lib/Transforms/Scalar/EarlyCSE.cpp:
```diff
@@ -553,6 +553,77 @@ bool DenseMapInfo<CallValue>::isEqual(CallValue LHS, CallValue RHS) {
+} // end namespace llvm
+
+unsigned DenseMapInfo<GEPValue>::getHashValue(const GEPValue &Val) {
+  GetElementPtrInst *GEP = cast<GetElementPtrInst>(Val.Inst);
```
nit: Use `auto *` with  `cast<>`.

dianqk committed:
fixup! [EarlyCSE] Compare GEP instructions based on offset

Use and and formatting

dianqk committed:
[EarlyCSE] Replace APInt with uint64_t

dianqk committed:
[EarlyCSE] Use std::optional<APInt>

DianQK commented:
> > > I'm wondering, would we be able to achieve similar results if we add a `DenseMap<GetElementPtrInst *, APInt>` cache for the offset but otherwise keep the previous code structure? Or would the extra hash lookup make this slow again?
> > 
> > 
> > I feel like I misunderstood. Do you mean putting the GEP back into `SimpleValue`? And then create a global (or EarlyCSE only) cache value?
> 
> Yes (EarlyCSE only cache). This is just a suggestion for something to check, but I suspect the performance of that approach will not be good enough and we have to go with this version.

I did not try this approach. My understanding is that this only adds extra lookups if applied to a single EarlyCSE. Since each result needs to be calculated once. The current patch is the no-hash approach as I understand it.
This result may be invalid if shared in EarlyCSE.

I made a new attempt to save uint64_t instead of APInt.
This result looks pretty good. https://llvm-compile-time-tracker.com/compare.php?from=ac1daad9bb4eb083df6b215c029816d3149e00d8&to=82f1eca7c180b0dfe8af36e933670f938a49bab1&stat=instructions%3Au

nikic commented on llvm/lib/Transforms/Scalar/EarlyCSE.cpp:
```diff
@@ -1561,6 +1646,31 @@ bool EarlyCSE::processNode(DomTreeNode *Node) {
+      auto *GEP = cast<GetElementPtrInst>(&Inst);
+      APInt Offset = APInt(SQ.DL.getIndexTypeSizeInBits(GEP->getType()), 0);
+      GEPValue GEPVal(GEP, GEP->accumulateConstantOffset(SQ.DL, Offset)
+                               ? Offset.tryZExtValue()
```
I think the offset should be `int64_t` and use `trySExtValue()` here. GEP offsets are signed. (Mainly for clarify, I think it will not make a functional difference?)

DianQK replied:
Yes, it's more clear to use `int64_t`.

nikic reviewed (approved):
LGTM

dianqk committed:
[EarlyCSE] Replace uint64_t with int64_t
